dipy/denoise/enhancement_kernel.pyx

```diff
@@ -0,0 +1,488 @@
+
+        with nogil:
+
+            for angv in range(OR1):
+                for angr in range(OR2):
+                    for xp in range(-hn, hn + 1):
+                        for yp in range(-hn, hn + 1):
+                            for zp in range(-hn, hn + 1):
+
+                                x[0] = xp
+                                x[1] = yp
+                                x[2] = zp
+
+                                lookuptablelocal[angv,
+                                                 angr,
+                                                 xp + hn,
+                                                 yp + hn,
+                                                 zp + hn] = self.k2(x, y, orientations[angr,:], orientations[angv,:])
+
+        # save to class member
+        self.lookuptable = lookuptablelocal
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    @cython.cdivision(True)
+    cdef void estimate_kernel_size(self, verbose=True):
+        """ Estimates the dimensions the kernel should
+        have based on the kernel parameters.
+
+        Parameters
+        ----------
+        verbose : boolean
+            Enable verbose mode.
+        """
+
+        cdef:
+            double [:] x
+            double [:] y
+            double [:] r
+            double [:] v
+            double i
+
+        x = np.array([0., 0., 0.])
+        y = np.array([0., 0., 0.])
+        r = np.array([0., 0., 1.])
+        v = np.array([0., 0., 1.])
+
+        # evaluate at origin
+        self.kernelmax = self.k2(x, y, r, v);
+
+        with nogil:
+
+            # determine a good kernel size
+            i = 0.0
+            while True:
+                i += 0.1
+                x[2] = i
+                kval = self.k2(x, y, r, v) / self.kernelmax
+                if(kval < 0.1):
+                    break;
+
+        N = ceil(i) * 2
+        if N % 2 == 0:
+            N -= 1
+
+        if verbose:
+            print("Dimensions of kernel: %dx%dx%d" % (N, N, N))
+
+        self.kernelsize = N
+
+    @cython.wraparound(False)
+    @cython.boundscheck(False)
+    @cython.nonecheck(False)
+    cdef double k2(self, double [:] x, double [:] y,
+                   double [:] r, double [:] v) nogil:
+        """ Evaluate the kernel at position x relative to
+        position y, with orientation r relative to orientation v.
+
+        Parameters
+        ----------
+        x : 1D ndarray
+            Position x
+        y : 1D ndarray
+            Position y
+        r : 1D ndarray
+            Orientation r
+        v : 1D ndarray
+            Orientation v
+
+        Returns
+        -------
+        kernel_value : double
+        """
+        cdef:
+            double [:] a
+            double [:,:] transm
+            double [:] arg1
+            double [:] arg2p
+            double [:] arg2
+            double [:] c
+            double kernelval
+
+        with gil:
```

Comment:
I see that you go back and forth with the gil stuff at a bunch of place, is it really needed? Seems like these four lines below should be fine, and the euler angle function seems simple enough also.